persistence/sql/persister_oauth2.go

```diff
@@ -491,18 +492,21 @@ func (p *Persister) GetRefreshTokenSession(ctx context.Context, signature string
 
 	var row OAuth2RefreshTable
 	if err := p.QueryWithNetwork(ctx).Where("signature = ?", signature).First(&row); errors.Is(err, sql.ErrNoRows) {
-		return nil, errorsx.WithStack(fosite.ErrNotFound)
+		return nil, errors.WithStack(fosite.ErrNotFound)
 	} else if err != nil {
 		return nil, sqlcon.HandleError(err)
 	}
 
-	gracePeriod := p.r.Config().RefreshTokenRotationGracePeriod(ctx)
 	if row.Active {
 		// Token is active
 		return row.toRequest(ctx, session, p)
-	} else if gracePeriod > 0 &&
+	}
+
+	if graceful := p.r.Config().GracefulRefreshTokenRotation(ctx); graceful.Period > 0 &&
 		row.FirstUsedAt.Valid &&
-		row.FirstUsedAt.Time.Add(gracePeriod).After(time.Now()) {
+		row.FirstUsedAt.Time.Add(graceful.Period).After(time.Now()) &&
+		(graceful.Count == 0 || // no limit
+			(row.UsedTimes.Valid && row.UsedTimes.Int32 < graceful.Count)) {
```

Comment:
I think we need to make explicit the definition of count in the naming, as I'm reading this here the counter starts at 0 and needs to be less than the configured value (that's fine) but it should be clear that the configured value is `<`

Reply:
This check runs before the counter is increased, hence setting the `count` config to 2 means you can reuse the token exactly twice (i.e. create two concurrent chains). If you have a better naming, happy to adjust.

Reply:
Ok, that's fine! Then we just need to document it that way :)